File: nlp/nlp.R
# Natural Language Processing - General model in R

# Importing the dataset
dataset_orig = read.delim('Restaurant_Reviews.tsv', quote = '', stringsAsFactors = FALSE)

# Cleaning the texts
#install.packages('tm')
#install.packages('SnowballC') # stopwords list
library(tm)
library(SnowballC)
corpus = VCorpus(VectorSource(dataset_orig$Review))
corpus = tm_map(corpus, content_transformer(tolower))
corpus = tm_map(corpus, removeNumbers)
corpus = tm_map(corpus, removePunctuation)
corpus = tm_map(corpus, removeWords, stopwords())
corpus = tm_map(corpus, stemDocument)
corpus = tm_map(corpus, stripWhitespace)

# Creating the Bag of Words model
dtm = DocumentTermMatrix(corpus)
# filter words in the sparse matrix - keep 99.9% columns remove non-frequent words
dtm = removeSparseTerms(dtm, 0.999)
dataset = as.data.frame(as.matrix(dtm))
# add category labels
dataset$Liked = dataset_orig$Liked

# Choose a classifier - Random Forest

# Encoding the target feature as factor
dataset$Liked = factor(dataset$Liked, levels = c(0, 1))

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Liked, SplitRatio = 0.80)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Fitting Random Forest to the Training set
#install.packages('randomForest')
library(randomForest) 
classifier = randomForest(x = training_set[-692],
                          y = training_set$Liked,
                          ntree = 10)

# Predicting the Test set results
y_pred = predict(classifier, newdata = test_set[-692])

# Confusion Matrix
cm = table(test_set[, 692], y_pred)

File: nlp/nlp.py
# Natural Language Processing - General model in Python
# Example data: Predicting restaurant reviews

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import re
import nltk
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
from sklearn.feature_extraction.text import CountVectorizer
#nltk.download('stopwords')

# Importing the dataset
dataset = pd.read_csv('Restaurant_Reviews.tsv', delimiter = '\t', quoting = 3)

# Cleaning the text
corpus = []
for i in range(0, 1000):
    # regex input - what types of characters we want to keep
    clean_review = re.sub('[^a-zA-Z]', ' ', dataset['Review'][i])
    # change all characters to lower case
    clean_review = clean_review.lower()
    clean_review = clean_review.split()
    # object to use the stemming class
    ps = PorterStemmer()
    # remove words that are in stopwords list
    # stem the words - use only roots
    clean_review = [ps.stem(word) for word in clean_review if not word in set(stopwords.words('English'))]
    clean_review = ' '.join(clean_review)
    corpus.append(clean_review)
    
# Creating the Bag of Words model
# keep only 1500 most common words
cv = CountVectorizer(max_features = 1500)
# create sparse matrix - reviews & words in corpus
X = cv.fit_transform(corpus).toarray()
# dependent var vector - category +/- 
Y = dataset.iloc[:, 1].values

# Choose a classifier - Naive Bayes  

# Split data into training set and test set
from sklearn.cross_validation import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y,
                                                    test_size = 0.20, random_state = 0)

# Fitting Naive Bayes to training set 
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, Y_train)

# Predicting the test set results
Y_pred = classifier.predict(X_test)

# Confusion matrix to evaluate performance
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Y_test, Y_pred)
